### ISSUE TITLE
fix software volume

### ISSUE THREAD

**Issue by Wouter1** [Wouter1/EMU-driver#27]
The software volume (eg, using the volume slider in OSX) is not working. EMU seems always at max volume regardless where you put the slider..

**Comment by Wouter1** [comment]
Checked the code. The code is configured for hard volume control - the master volume is sent directly to the EMU. Maybe there is something broken there. I'm not sure how this is supposed to work, as there also is a hard volume slider on the EMU. Would the software selected volume be multiplied with the hard-set volume on the EMU?

There is also code "addSoftVolumeControls" that controls volume from the software, but this is disabled. 

There is a field mOutputVolume in the EMUUSBAudioEngine but that's disabled as addSoftVolumeControls is not called.

**Comment by Wouter1** [comment]
Checked the windows driver CAudioDevice::SetMasterVolume, apparently it uses software volume anyway. 

There is also some volume related code in Unit.cpp. Here something seems to be happening with USB volume request. There seems a check there  if there is volume support
```
			if (m_ParameterBlock[ChannelNumber].Volume.Support)
```

**Comment by Wouter1** [comment]
The AK4396 has a volume controller (linear 256 steps) in the hardware

**Comment by Wouter1** [comment]
linear 256 steps is useless anyway.

**Comment by Wouter1** [comment]
Just enabling ENABLESOFTCONTROLS in EMUUSBAudioDevice.cpp (why isn't it in .h file??) seems enough. 
I'll check the code to see what else happens when I do that

**Comment by Wouter1** [comment]
There is also a INPUT volume controller created

And a MUTE control for both input and output.

**Comment by Wouter1** [comment]
Checked mute with HALLab tool, seems ok

Checked mute with sound control panel, seems ok too.

**Comment by Wouter1** [comment]
CHecked volume input control with audacity input volume slider. Works ok

Checked input volume control with control panel. Works ok.

Checked input mute with HALLab tool. Works ok.

**Comment by Wouter1** [comment]
Max volume seems 7200. 

```
EMUUSBAudioEngine::softwareVolumeChangedHandler input: channel= 0 oldValue= 7109 newValue= 7200
``` 

Not clear what that means, but we must have no scaling when max volume is selected (minimum amount of processing when max volume).

**Comment by Wouter1** [comment]
Checked output volume. It is correctly going to 1.000000000 (checked 9 decimals) at max output volume (slider to max)

**Comment by Wouter1** [comment]
Need to document too that recommended is to leave this at maximum and use potmeters on the EMU

**Comment by Wouter1** [comment]
Added check in input volume scaler to return immediately instead of multiplying all input values with 1.

**Comment by Wouter1** [comment]
That should also fix the output as that uses the same EMUUSBAudioClip#Volume() function

**Comment by Wouter1** [comment]
bump version to 3.5.5

**Comment by Wouter1** [comment]
release test needed.

**Comment by Wouter1** [comment]
Maybe I can make a topic how to make the highest quality recording
 * volumes to 1.0 (clipping can never be avoided with soft volume, you only loose resolution)
 * 96kHz (192k gives extra noise it seems, double check)
 * connect as straight as possible, no long cables, keep cables away from EM noise
 * consider the mods if you really need high end

**Comment by Wouter1** [comment]
Updated the test procedure: use volume around 0.9 for input and output.
This is to ensure the volume multiplier is actually working, which will take extra CPU and therefore needs to be enabled

**Comment by Wouter1** [comment]
| rate | no clicking |
| --- | --- |
| 44k | 2 clicks |
| 44k | ok |
| 44k | ok |
| 44k | ok |
| 96k | ok |

**Comment by Wouter1** [comment]
Had 2 clicks at 44k. Seemed a one-time off. Maybe because I let the screen go to sleep while testing.

**Comment by Wouter1** [comment]
approved.

**Comment by Wouter1** [comment]
Released, v3.5.5.

the 'how to make quality recording' is now ticket #61.
